libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -1564,17 +1629,103 @@ bool AP_Mission::advance_current_nav_cmd(uint16_t starting_index)
+                                }
+                                _translation.calculated = true; // do that just once at very first WayPoint (Start-Waypoint)
+                                _translation.lat = _start_loc.lat - cmd.content.location.lat;
+                                _translation.lng = _start_loc.lng - cmd.content.location.lng;
+
+                                // calculate altitude-translation
+                                if (cmd.content.location.relative_alt == 1) {
+                                    _translation.alt = _start_loc.alt - cmd.content.location.alt - _translation.alt; // subtract Home-Altitude if WP-altitude is relative to
+                                } else {
+                                    _translation.alt = _start_loc.alt - cmd.content.location.alt;
+                                }
+
+                                Mission_Command tmp;
+                                // get direction from Homepoint to Start-Waypoint
+                                tmp.content.location = AP::ahrs().get_home();
+                                _translation.direction = tmp.content.location.get_bearing_to(_start_loc);   // in centidegrees from 0 36000
+
+                                // direction from Homepoint to not translated Base-Waypoint / amount of rotation (relative to Base-Waypoint)
+                                _translation.direction -= tmp.content.location.get_bearing_to(cmd.content.location);
+                                if (_translation.direction < 0) {
+                                    _translation.direction += 36000;
+                                }
+
+                                cmd.content.location = _start_loc;
+                                switch (restart_behaviour) {
+                                case Restart_Behaviour::RESTART_PARALLEL_TRANSLATED:
+                                    gcs().send_text(MAV_SEVERITY_INFO, "Mission: %u %s Restart Parallel Translated", cmd.index, cmd.type());
+                                    break;
+                                case Restart_Behaviour::RESTART_ROTATED_TRANSLATED:
+                                    gcs().send_text(MAV_SEVERITY_INFO, "Mission: %u %s Restart Rotated", cmd.index, cmd.type());
+                                    break;
+                                default:
+                                    gcs().send_text(MAV_SEVERITY_NOTICE, "Mission: %u %s Restart Mode unknown %i", cmd.index, cmd.type(), static_cast<int>(restart_behaviour));
+                                }
+                                gcs().send_text(MAV_SEVERITY_INFO, "Mission: alt-transmission %.2fm",  static_cast<float>(_translation.alt)/100.0);
+                            }
+                            else {
+                                if (restart_behaviour >= Restart_Behaviour::RESTART_PARALLEL_TRANSLATED){ // do at least parallel translation
+                                    cmd.content.location.lat += _translation.lat;
```

Comment:
This is not accurately translating anything. You can't simply add lat/long pairs, you actually have to project it down a bearing at a distance. It probably looks correct at a small scale, but it will slowly be warping the mission.